### PR TITLE
[SYCL] Implement sycl_ext_oneapi_get_kernel_info extension

### PR DESCRIPTION
**Pull request by againull**
Extension: [sycl_ext_oneapi_get_kernel_info](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_get_kernel_info.asciidoc)

Outlined kernel::attributes test to a separate file with existing XFAIL markings as it is the only descriptor which has a problem (GSD-8971) and reenabled kernel_info.cpp for other descriptors.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/get_kernel_info.hpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+typename sycl::detail::is_kernel_device_specific_info_desc<Param>::return_type
+get_kernel_info(const context &Ctx, const device &Dev) {
+  auto Bundle =
+      sycl::get_kernel_bundle<KernelName, sycl::bundle_state::executable>(Ctx);
+  return Bundle.template get_kernel<KernelName>().template get_info<Param>(Dev);
+}
+
+template <typename KernelName, typename Param>
+typename sycl::detail::is_kernel_device_specific_info_desc<Param>::return_type
+get_kernel_info(const queue &Q) {
```
Comment (uditagarwal97) [inline sycl/include/sycl/ext/oneapi/get_kernel_info.hpp:38]: Can we extend of the test cases to also cover this overload of `get_kernel_info`? Currently, the tests only cover the other two overloads.

Comment (againull) [reply sycl/include/sycl/ext/oneapi/get_kernel_info.hpp:38]: Good point, will do.

File: sycl/test-e2e/Basic/kernel_info.cpp [hunk 2]
```diff
@@ -82,12 +89,70 @@ int main() {
...
+              dev);
+      // Use ext_oneapi_get_kernel_info extension and check that answers match.
+      auto globalWorkSizeExt = syclex::get_kernel_info<
+          SingleTask, info::kernel_device_specific::global_work_size>(ctx, dev);
+      assert(globalWorkSize == globalWorkSizeExt);
+      // Use ext_oneapi_get_kernel_info extension with queue parameter and check
+      // the result.
+      auto globalWorkSizeExtQ = syclex::get_kernel_info<
+          SingleTask, info::kernel_device_specific::global_work_size>(q);
+      assert(globalWorkSize == globalWorkSizeExtQ);
```
Comment (gmlueck) [inline sycl/test-e2e/Basic/kernel_info.cpp:155]: Same comment here about exceptions.  The call to `krn.get_info<sycl::info::kernel_device_specific::global_work_size>()` throws an exception, so this test never even makes the two calls to `syclex::get_kernel_info`.

File: sycl/test-e2e/Basic/kernel_info.cpp [hunk 3]
```diff
@@ -59,8 +58,16 @@ int main() {
...
-  assert(krnAttr.empty());
+
+  // Use ext_oneapi_get_kernel_info extension and check that answers match.
+  const context krnCtxExt =
+      syclex::get_kernel_info<SingleTask, info::kernel::context>(ctx);
+  assert(krnCtxExt == krnCtx);
+  // Reference count might be different because we have to retain the kernel
+  // handle first to fetch the info. So just check that it is not 0.
+  const cl_uint krnRefCountExt =
+      syclex::get_kernel_info<SingleTask, info::kernel::reference_count>(ctx);
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/test-e2e/Basic/kernel_info.cpp:69]: Created: CMPLRLLVM-62644

File: sycl/test-e2e/Basic/kernel_info.cpp [hunk 4]
```diff
@@ -42,6 +37,10 @@ int main() {
   bool ExceptionWasThrown = false;
   try {
     const cl_uint krnArgCount = krn.get_info<info::kernel::num_args>();
+    // Use ext_oneapi_get_kernel_info extension and check that answers match.
+    const cl_uint krnArgCountExt =
+        syclex::get_kernel_info<SingleTask, info::kernel::num_args>(ctx);
+    assert(krnArgCountExt == krnArgCount);
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/test-e2e/Basic/kernel_info.cpp]: Missed that, thank you, fixed.

File: sycl/test-e2e/Basic/kernel_info.cpp (hunk 2 above)
Comment (againull) [reply sycl/test-e2e/Basic/kernel_info.cpp]: Fixed.

**Comment by againull** [comment]
@gmlueck  Could you please take another look, thanks in advance!

**Comment by sarnex** [comment]
@againull Seeing this fail in postcommit, can you please take a look? Thanks
https://github.com/intel/llvm/actions/runs/11350085536/job/31568113561
```
-- Testing: 2212 tests, 24 workers --
FAIL: SYCL :: no-xfail-without-tracker.cpp (2121 of 2212)
******************** TEST 'SYCL :: no-xfail-without-tracker.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 26
grep -rI "XFAIL:" /__w/llvm/llvm/llvm/sycl/test-e2e -A 1 --include=*.c --include=*.cpp      --no-group-separator |  grep -v "XFAIL:" |  grep -Pv "XFAIL-TRACKER:\s+(?:[https://github.com/[\w\d-]+/[\w\d-]+/issues/[\d]+)](https://github.com/[/w/d-]+/[/w/d-]+/issues/[/d]+))|(?:[\w]+-[\d]+)" |  wc -l | /__w/llvm/llvm/toolchain/bin/FileCheck /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp --check-prefix NUMBER-OF-XFAIL-WITHOUT-TRACKER
# executed command: grep -rI XFAIL: /__w/llvm/llvm/llvm/sycl/test-e2e -A 1 '--include=*.c' '--include=*.cpp' --no-group-separator
# note: command had no output on stdout or stderr
# executed command: grep -v XFAIL:
# note: command had no output on stdout or stderr
# executed command: grep -Pv 'XFAIL-TRACKER:\s+(?:[https://github.com/[\w\d-]+/[\w\d-]+/issues/[\d]+)](https://github.com/[/w/d-]+/[/w/d-]+/issues/[/d]+))|(?:[\w]+-[\d]+)'
# note: command had no output on stdout or stderr
# executed command: wc -l
# note: command had no output on stdout or stderr
# executed command: /__w/llvm/llvm/toolchain/bin/FileCheck /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp --check-prefix NUMBER-OF-XFAIL-WITHOUT-TRACKER
# .---command stderr------------
# | /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp:50:37: error: NUMBER-OF-XFAIL-WITHOUT-TRACKER: expected string not found in input
# | // NUMBER-OF-XFAIL-WITHOUT-TRACKER: 179
# |                                     ^
# | <stdin>:1:1: note: scanning from here
# | 178
# | ^
# | 
# | Input file: <stdin>
# | Check file: /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp
# | 
# | -dump-input=help explains the following input dump.
# | 
# | Input was:
# | <<<<<<
# |           1: 178 
# | check:50     X~~~ error: no match found
# | >>>>>>
# `-----------------------------
# error: command failed with exit status: 1
```

**Comment by againull** [comment]
> @againull Seeing this fail in postcommit, can you please take a look? Thanks https://github.com/intel/llvm/actions/runs/11350085536/job/31568113561
> 
> ```
> -- Testing: 2212 tests, 24 workers --
> FAIL: SYCL :: no-xfail-without-tracker.cpp (2121 of 2212)
> ******************** TEST 'SYCL :: no-xfail-without-tracker.cpp' FAILED ********************
> Exit Code: 1
> 
> Command Output (stdout):
> --
> # RUN: at line 26
> grep -rI "XFAIL:" /__w/llvm/llvm/llvm/sycl/test-e2e -A 1 --include=*.c --include=*.cpp      --no-group-separator |  grep -v "XFAIL:" |  grep -Pv "XFAIL-TRACKER:\s+(?:[https://github.com/[\w\d-]+/[\w\d-]+/issues/[\d]+)](https://github.com/[/w/d-]+/[/w/d-]+/issues/[/d]+))|(?:[\w]+-[\d]+)" |  wc -l | /__w/llvm/llvm/toolchain/bin/FileCheck /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp --check-prefix NUMBER-OF-XFAIL-WITHOUT-TRACKER
> # executed command: grep -rI XFAIL: /__w/llvm/llvm/llvm/sycl/test-e2e -A 1 '--include=*.c' '--include=*.cpp' --no-group-separator
> # note: command had no output on stdout or stderr
> # executed command: grep -v XFAIL:
> # note: command had no output on stdout or stderr
> # executed command: grep -Pv 'XFAIL-TRACKER:\s+(?:[https://github.com/[\w\d-]+/[\w\d-]+/issues/[\d]+)](https://github.com/[/w/d-]+/[/w/d-]+/issues/[/d]+))|(?:[\w]+-[\d]+)'
> # note: command had no output on stdout or stderr
> # executed command: wc -l
> # note: command had no output on stdout or stderr
> # executed command: /__w/llvm/llvm/toolchain/bin/FileCheck /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp --check-prefix NUMBER-OF-XFAIL-WITHOUT-TRACKER
> # .---command stderr------------
> # | /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp:50:37: error: NUMBER-OF-XFAIL-WITHOUT-TRACKER: expected string not found in input
> # | // NUMBER-OF-XFAIL-WITHOUT-TRACKER: 179
> # |                                     ^
> # | <stdin>:1:1: note: scanning from here
> # | 178
> # | ^
> # | 
> # | Input file: <stdin>
> # | Check file: /__w/llvm/llvm/llvm/sycl/test-e2e/no-xfail-without-tracker.cpp
> # | 
> # | -dump-input=help explains the following input dump.
> # | 
> # | Input was:
> # | <<<<<<
> # |           1: 178 
> # | check:50     X~~~ error: no match found
> # | >>>>>>
> # `-----------------------------
> # error: command failed with exit status: 1
> ```


Yes, here is the fix: https://github.com/intel/llvm/pull/15709

**Pull request merged**